### PR TITLE
Fail if inputs contain nested nulls in set_agg Presto aggregates

### PR DESCRIPTION
**Pull request by duanmeng**
In Presto, set_agg checks for null array elements, map values,
and struct fields and throws, but in Velox, they don't. For example,
the following SQL should throw an exception, "ARRAY comparison
not supported for arrays with null elements".

```SQL
presto> SELECT set_agg(col1) FROM (VALUES (1, array [1, 2, 3]), 
(2, array [null, 3, 1]), (2, array [null, 3, 1])) AS tbl(col0, col1);
Query 20230918_072245_00001_fggjk failed: ARRAY 
comparison not supported for arrays with null elements
```
This PR do the following things,
- Add a check that all raw inputs (in the `Partial/Single` step) 
of set_agg do not contain nested nulls before serializing them to the accumulator.
- **Fix a bug** of checkNulls in min/max that uses the wrong index. It must use
the base index, not the top-level index.

Part of https://github.com/facebookincubator/velox/issues/6314

### CODE REVIEW

File: velox/functions/prestosql/aggregates/SetAggregates.cpp [hunk 1]
```diff
@@ -199,8 +199,11 @@ class SetBaseAggregate : public exec::Aggregate {
 template <typename T>
 class SetAggAggregate : public SetBaseAggregate<T> {
  public:
+  SetAggAggregate(const TypePtr& resultType, const bool throwOnNestedNulls)
+      : SetBaseAggregate<T>(resultType),
+        throwOnNestedNulls_(throwOnNestedNulls) {}
   explicit SetAggAggregate(const TypePtr& resultType)
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/SetAggregates.cpp:205]: add empty line

Do we need 2 constructors? Can we use one?

File: velox/functions/prestosql/aggregates/SetAggregates.cpp [hunk 2]
```diff
@@ -274,9 +287,32 @@ class SetAggAggregate : public SetBaseAggregate<T> {
...
+      if (throwOnNestedNulls_) {
+        checkNulls(Base::decoded_, i);
+      }
+
       accumulator->addValue(Base::decoded_, i, Base::allocator_);
     });
   }
+
+ private:
+  bool checkNulls(const DecodedVector& decoded, vector_size_t index) const {
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/SetAggregates.cpp:299]: return value is not used; change to void

File: velox/functions/prestosql/aggregates/SetAggregates.cpp [hunk 3]
```diff
@@ -274,9 +287,32 @@ class SetAggAggregate : public SetBaseAggregate<T> {
...
+    if (decoded.isNullAt(index)) {
+      return true;
+    }
+
+    if (throwOnNestedNulls_) {
+      VELOX_USER_CHECK(
+          !decoded.base()->containsNullAt(index),
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/SetAggregates.cpp:306]: This is incorrect. We should use decoded.index(index)

```
!decoded.base()->containsNullAt(decoded.index(index)),
```

Same problem exists in min/max aggregate function.

Comment (duanmeng) [reply velox/functions/prestosql/aggregates/SetAggregates.cpp]: Got it, must use the base index! I've fixed it along with the min/max aggregate function.

File: velox/docs/functions/presto/aggregate.rst [hunk 4]
```diff
@@ -193,6 +193,7 @@ General Aggregate Functions
 .. function:: set_agg(x) -> array<[same as input]>
 
     Returns an array created from the distinct input ``x`` elements.
+    ``x`` must not contains nulls when it is complex type.
```
[the remark this answers is not in the document]
Comment (duanmeng) [reply velox/docs/functions/presto/aggregate.rst]: Updated, and min/max too.

File: velox/functions/prestosql/aggregates/SetAggregates.cpp (hunk 1 above)
Comment (duanmeng) [reply velox/functions/prestosql/aggregates/SetAggregates.cpp]: Fixed

File: velox/functions/prestosql/aggregates/SetAggregates.cpp (hunk 2 above)
Comment (duanmeng) [reply velox/functions/prestosql/aggregates/SetAggregates.cpp]: Fixed

**Review by mbasmanova — COMMENTED** [review]
Thanks. Perhaps, mention bug fix for min/max in the PR description.

File: velox/functions/prestosql/aggregates/SetAggregates.cpp [hunk 5]
```diff
@@ -273,10 +287,35 @@ class SetAggAggregate : public SetBaseAggregate<T> {
...
+      const vector_size_t* indices,
+      vector_size_t index) const {
+    if (decoded.isNullAt(index)) {
+      return;
+    }
+
+    if (throwOnNestedNulls_) {
+      VELOX_USER_CHECK(
+          !decoded.base()->containsNullAt(indices[index]),
+          fmt::format(
```
[the remark this answers is not in the document]
Comment (duanmeng) [reply velox/functions/prestosql/aggregates/SetAggregates.cpp]: Removed

File: velox/functions/prestosql/aggregates/SetAggregates.cpp [hunk 6]
```diff
@@ -214,6 +217,13 @@ class SetAggAggregate : public SetBaseAggregate<T> {
       VectorPtr& result) const override {
     const auto& elements = args[0];
 
+    if (throwOnNestedNulls_) {
+      DecodedVector decodedElements(*elements, rows);
```
[the remark this answers is not in the document]
Comment (duanmeng) [reply velox/functions/prestosql/aggregates/SetAggregates.cpp:221]: @mbasmanova We could not use `Base::decoded_` here for `toIntermediate` is marked as const while `Base::decoded_.indices()` is not.

**Pull request closed without merging**